tests/backends/tango/test_dsr.py

```diff
@@ -1,37 +1,125 @@
+
+    @pytest.fixture(scope="class")
+    def client_write(self, tango_context):
+        def _write_attribute(path: list[str], expected: Any):
+            attribute = "_".join(path)
+            with self.controller.assertPerformed(pascal_2_snake(path), "WRITE"):
+                tango_context.write_attribute(attribute, expected)
+
+        return _write_attribute
+
+    @pytest.fixture(scope="class")
+    def client_exec(self, tango_context):
+        def _exec_command(path: list[str]):
+            command = "_".join(path)
+            with self.controller.assertPerformed(pascal_2_snake(path), "EXECUTE"):
+                tango_context.command_inout(command)
+
+        return _exec_command
+
+    def test_list_attributes(self, tango_context):
+        assert list(tango_context.get_attribute_list()) == [
+            "BigEnum",
+            "ReadBool",
+            "ReadInt",
+            "ReadWriteFloat",
+            "ReadWriteInt",
+            "StringEnum",
+            "WriteBool",
+            "SubController01_ReadInt",
+            "SubController02_ReadInt",
+            "State",
+            "Status",
+        ]
+
+    def test_list_commands(self, tango_context):
+        assert list(tango_context.get_command_list()) == [
+            "Go",
+            "Init",
+            "State",
+            "Status",
+        ]
+
+    def test_state(self, tango_context):
+        assert tango_context.command_inout("State") == DevState.ON
+
+    def test_status(self, tango_context):
+        expected = "The device is in ON state."
+        assert tango_context.command_inout("Status") == expected
+
+    def test_read_int(self, client_read):
+        client_read(["ReadInt"], AttrR(Int())._value)
+
+    def test_read_write_int(self, client_read, client_write):
+        client_read(["ReadWriteInt"], AttrR(Int())._value)
+        client_write(["ReadWriteInt"], AttrR(Int())._value)
+
+    def test_read_write_float(self, client_read, client_write):
+        client_read(["ReadWriteFloat"], AttrR(Float())._value)
+        client_write(["ReadWriteFloat"], AttrR(Float())._value)
+
+    def test_read_bool(self, client_read):
+        client_read(["ReadBool"], AttrR(Bool())._value)
+
+    def test_write_bool(self, client_write):
+        client_write(["WriteBool"], AttrR(Bool())._value)
+
+    def test_string_enum(self, client_read, client_write):
+        enum = AttrR(String(), allowed_values=["red", "green", "blue"])._value
+        client_read(["StringEnum"], enum)
+        client_write(["StringEnum"], enum)
+
+    def test_big_enum(self, client_read):
+        client_read(["BigEnum"], AttrR(Int(), allowed_values=list(range(1, 18)))._value)
+
+    def test_go(self, client_exec):
+        client_exec(["Go"])
+
+    def test_read_child1(self, client_read):
+        client_read(["SubController01", "ReadInt"], AttrR(Int())._value)
+
+    def test_read_child2(self, client_read):
+        client_read(["SubController02", "ReadInt"], AttrR(Int())._value)
```

Comment:
On as second look, I think these tests are suffering in readability by trying to be too DRY. It isn't really clear what is being tested and I think someone trying to add new tests will either struggle to understand how to do it, or worse just copy and paste things without understanding what they are testing.

The `tango_context` fixture is good, but I think it should be used directly within each test. I think AssertableController is too much. Having a utility to recurse down the subcontrollers is useful, but I think the mocking and assertions should be done in each test.

What do you think? Maybe we could pair on this and see if we can improve it?

Reply:
My argument is that I have been able to apply a consistent approach to other "backends", I think consistency between different backends is favorable so that we dont have to maintain bespoke things for every possible variation.  Alternatively, someone adding tests will have to understand multiple different ways. If I needed to add soft ioc tests today: I dont understand how to do it and would have to copy and paste things without understanding what I am testing because there is no structure. Also I have 600 lines to search to copy and paste from and need to have in depth knowledge of the backend to mock the right things

Happy to pair and discuss - Its not clear to me why you would divert from having a consistent way of testing

Reply:
I think this looks much clearer now!